### PR TITLE
Sophiex/dev/latent noise

### LINKED ISSUE #593: Latent noise interpolation
### Is your feature request related to a problem? Please describe.

We want to do masked token modelling pre-training with interpolated latent noise following https://arxiv.org/abs/2507.15856

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

Draft for how to implement interpolated noise during masked token modelling pre-training.

Closes #593 

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

File: config/default_config.yml [hunk 1]
```diff
@@ -136,4 +142,4 @@ run_id: ???
 # Parameters for logging/printing in the training loop
 train_log:
   # The period to log metrics (in number of batch steps)
-  log_interval: 20
```
Comment (shmh40) [inline config/default_config.yml]: I never understand this one

File: config/default_config.yml [hunk 2]
```diff
@@ -17,13 +18,13 @@ ae_local_with_qk_lnorm: True
 
 ae_local_num_queries: 1
 ae_local_queries_per_cell: False
-ae_adapter_num_heads: 16
+ae_adapter_num_heads: 8
```
Comment (shmh40) [inline config/default_config.yml]: Usual comment for default_config

File: config/default_config.yml [hunk 3]
```diff
@@ -61,6 +62,11 @@ attention_dtype: bf16
 mlp_norm_eps: 1e-5
 norm_eps: 1e-4
 
+kl_weight: 0.0 # 1e-5
+noise_gamma: 2.0
+use_additive_noise: False
+deterministic_latents: False
+
```
Comment (shmh40) [inline config/default_config.yml:69]: Looks good

File: config/default_config_atos.yml [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+embed_local_coords: True
+embed_centroids_local_coords: False
+embed_size_centroids: 0
+embed_unembed_mode: "block"
+
+target_cell_local_prediction: True
+
+ae_local_dim_embed: 1024
+ae_local_num_blocks: 2
+ae_local_num_heads: 16
```
Comment (shmh40) [inline config/default_config_atos.yml:16]: Remove

File: config/default_config_leonardo.yml [hunk 5]
```diff
@@ -0,0 +1,129 @@
+streams_directory: "./config/streams/streams_anemoi/"
```
Comment (shmh40) [inline config/default_config_leonardo.yml]: "

File: config/default_config_atos.yml [hunk 6]
```diff
@@ -0,0 +1,129 @@
+streams_directory: "./config/streams/streams_anemoi/"
```
Comment (shmh40) [inline config/default_config_atos.yml]: Usual, remove before merge

File: config/streams/streams_anemoi/era5.yml [hunk 7]
```diff
@@ -11,7 +11,7 @@ ERA5 :
   type : anemoi
```
Comment (shmh40) [inline config/streams/streams_anemoi/era5.yml]: "

File: src/weathergen/model/attention.py [hunk 8]
```diff
@@ -153,7 +154,8 @@ def sparsity_mask(score, b, h, q_idx, kv_idx):
 
             return flex_attention(qs, ks, vs, score_mod=sparsity_mask)
 
-        self.compiled_flex_attention = torch.compile(att, dynamic=False)
+        self.compiled_flex_attention = torch.compile(att, dynamic=False, mode="max-autotune")
```
Comment (shmh40) [inline src/weathergen/model/attention.py]: Should these changes be in a different PR? What is happening here - was this to make the flex attention work on Leonardo or whatever?

File: src/weathergen/model/model.py [hunk 9]
```diff
@@ -202,6 +203,14 @@ def create(self) -> "Model":
         # local assimilation engine
         self.ae_local_blocks = LocalAssimilationEngine(cf).create()
 
+        if cf.kl_weight > 0.0:
```
Comment (shmh40) [inline src/weathergen/model/model.py]: So the entry point to use the latent noise is kl_weight > 0.0 in the config. I am happy with this, but do you think it would be clearer to people to have a True/False in the config? I think not necessary for now, but just mentioning for future.

use_latent_noise : true
kl_weight : 1e-5
noise_gamme : whatever

Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: I wanted to avoid yet another config value, but happy to change that. What do you think/prefer?

File: src/weathergen/model/attention.py (hunk 8 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/attention.py]: Will revert, this was me debugging things

File: src/weathergen/model/model.py [hunk 10]
```diff
@@ -356,7 +365,7 @@ def create(self) -> "Model":
             )
             self.pred_heads.append(
                 EnsPredictionHead(
-                    dims_embed[-1],
+                    dims_embed[0],
```
Comment (shmh40) [inline src/weathergen/model/model.py]: This change makes no difference right?

File: src/weathergen/model/model.py [hunk 11]
```diff
@@ -613,8 +622,15 @@ def assimilate_local(
         )
 
         # # local assimilation model
-        # for block in self.ae_local_blocks :
-        #   tokens = checkpoint( block, tokens, cell_lens, use_reentrant=False)
+        # for block in self.ae_local_blocks:
```
Comment (shmh40) [inline src/weathergen/model/model.py:636]: Remove

File: src/weathergen/model/parametrised_prob_dist.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        self.mean_and_var = nn.Linear(dim, 2 * dim, bias=False)
+
+    def interpolate_with_noise(self, z, sampling=False, noise_level=-1):
+        self.diag_gaussian.reset_parameters(self.mean_and_var(z))
+        z_latents = self.diag_gaussian.sample() if sampling else self.diag_gaussian.mean
+
+        if self.training and self.gamma > 0.0:
+            device = z_latents.device
+            s = z_latents.shape
+            bsz = 1
```
Comment (shmh40) [inline src/weathergen/model/parametrised_prob_dist.py]: Sorry, can you call this batch_size. 

And can we just get the batch_size here straight from the default config to address your TODO below?

Comment (shmh40) [reply src/weathergen/model/parametrised_prob_dist.py]: Otherwise I think this looks good

File: src/weathergen/train/trainer.py [hunk 13]
```diff
@@ -50,7 +51,18 @@ def init(
         self,
         cf: Config,
     ):
-        self.cf = cf
+        self.cf = OmegaConf.merge(
```
Comment (shmh40) [inline src/weathergen/train/trainer.py:56]: Sorry, maybe I am being stupid but why are we doing this? Can't this stuff come directly from the config now anyway?

Comment (shmh40) [reply src/weathergen/train/trainer.py:56]: I guess it is left in to make sure it still works with old configs?

File: src/weathergen/model/attention.py [hunk 14]
```diff
@@ -94,7 +94,8 @@ def forward(self, x, x_lens, ada_ln_aux=None):
             dropout_p=self.dropout_rate,
         )
 
-        out = self.proj_out(outs.flatten(-2, -1))
+        out = self.dropout(self.proj_out(outs.flatten(-2, -1)))
```
Comment (shmh40) [inline src/weathergen/model/attention.py]: Another PR?

File: src/weathergen/model/model.py [hunk 15]
```diff
@@ -652,6 +669,15 @@ def assimilate_local(
             for block in self.ae_local_blocks:
                 tokens_c = checkpoint(block, tokens_c, cell_lens_c, use_reentrant=False)
 
+            if self.cf.kl_weight > 0.0:
```
Comment (shmh40) [inline src/weathergen/model/model.py]: Despite what I said before, I am actually not sure where this should happen. I have the sense that doing it for the global latents has some neat properties, but it could be local, or both. Do you have strong views on this?

File: src/weathergen/train/trainer.py [hunk 16]
```diff
@@ -466,11 +481,16 @@ def train(self, epoch):
                 dtype=self.mixed_precision_dtype,
                 enabled=cf.with_mixed_precision,
             ):
-                preds = self.ddp_model(self.model_params, batch, cf.forecast_offset, forecast_steps)
+                preds, posteriors = self.ddp_model(
```
Comment (shmh40) [inline src/weathergen/train/trainer.py:503]: It should probably happen in another PR, but can we start to think about what we need to start generating an ensemble with the sampled latents and start training with CRPS? 

We need to sample the differently noised latents, and pass each one through the global and then the decoders, and then store with [sample_size, preds], then in the loss all these samples are used to compute the CRPS against the true target. I don't think is very difficult? Can you think of where this would be tricky? It is slightly different task, as this should encourage us to learn an appropriate std. dev. for the noised latents, rather than focussing on just making the task harder, but I think the two could work in tandem (although perhaps this is not as neat as it should be, perhaps there should be a separation of the tasks).

File: src/weathergen/model/model.py (hunk 10 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: -1 is more correct

File: src/weathergen/model/model.py (hunk 15 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py]: Yea happy to discuss this, I think more broadly, I want a clearer delineation between encoder and decoder. Conceptually, to me the encoder should not write to the masked tokens

File: src/weathergen/model/model.py (hunk 9 above)
Comment (shmh40) [reply src/weathergen/model/model.py]: I think it is fine for now, likely just a few users for the next few weeks and you are right about config clutter!

File: src/weathergen/model/parametrised_prob_dist.py (hunk 12 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/parametrised_prob_dist.py]: I think so much of the code relies on batchsize 1 in a number of places and I don't think this will change in the near future, given we are planning for more datasets not fewer, that it isn't worth the time, especially given that I don't think the rest of the code works with batchsize >1

File: src/weathergen/train/trainer.py (hunk 13 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py:56]: yes backwards compatibility

File: src/weathergen/train/trainer.py (hunk 16 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py:503]: I think we (in this project) conceptually muddy a few concepts. The pre-training of latents is different from an autoencoder which again differs from forecasting, I think it is worth having a longer discussion about what we want to learn when.

Comment (shmh40) [reply src/weathergen/train/trainer.py:503]: Agreed, sounds good

File: src/weathergen/model/model.py (hunk 11 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/model.py:636]: I believe this has always been there for documentation purposes

Comment (clessig) [reply src/weathergen/model/model.py:636]: yes

**Review by shmh40 — COMMENTED** [review]
Looks good to me if configs are reverted as expected, previously tested on Leonardo. I have not evaluated the performance through to scores though.

File: config/default_config.yml [hunk 17]
```diff
@@ -17,7 +17,7 @@ ae_local_with_qk_lnorm: True
 
 ae_local_num_queries: 1
 ae_local_queries_per_cell: False
-ae_adapter_num_heads: 8
+ae_adapter_num_heads: 16
```
Comment (clessig) [inline config/default_config.yml]: Is this change to the default config necessary?

File: config/default_config.yml [hunk 18]
```diff
@@ -23,7 +23,7 @@ ae_adapter_with_qk_lnorm: True
 ae_adapter_with_residual: True
 ae_adapter_dropout_rate: 0.1
 
-ae_global_dim_embed: 2048
+ae_global_dim_embed: 2048
```
Comment (clessig) [inline config/default_config.yml]: Is this ruffed? There's an extra whitespace

File: config/default_config.yml [hunk 19]
```diff
@@ -61,6 +61,11 @@ attention_dtype: bf16
 mlp_norm_eps: 1e-5
 norm_eps: 1e-4
 
+kl_weight: 0.0 # 1e-5
+noise_gamma: 2.0
```
Comment (clessig) [inline config/default_config.yml]: Could we add a common prefix to these parameters so that it's clear which part of the model/training they affect (so something like `ae_global_?`

File: config/streams/era5_1deg/era5.yml [hunk 20]
```diff
@@ -30,7 +30,7 @@ ERA5 :
   target_readout :
     type : 'obs_value'  # token or obs_value
     num_layers : 2
-    num_heads : 4
+    num_heads : 8
```
Comment (clessig) [inline config/streams/era5_1deg/era5.yml]: Does this change really fit this PR?

File: src/weathergen/model/attention.py [hunk 21]
```diff
@@ -154,6 +155,7 @@ def sparsity_mask(score, b, h, q_idx, kv_idx):
             return flex_attention(qs, ks, vs, score_mod=sparsity_mask)
 
         self.compiled_flex_attention = torch.compile(att, dynamic=False)
+        # self.compiled_flex_attention = flex_attention
```
Comment (clessig) [inline src/weathergen/model/attention.py]: Why do the lines appear here as new? In any case, I think they should be removed.

File: src/weathergen/model/parametrised_prob_dist.py [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+        return 0.5 * torch.sum(
+            logtwopi + self.logvar + torch.pow(sample - self.mean, 2) / self.var,
+            dim=dims or self.sum_dims,
+        )
+
+    def mode(self):
+        return self.mean
+
+
+class InterpolatedLatents(nn.Module):
```
Comment (clessig) [inline src/weathergen/model/parametrised_prob_dist.py]: The class implements/performs the interpolation so `LatentInterpolator` seems like a better name?

File: src/weathergen/model/parametrised_prob_dist.py [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+        self.mean_and_var = nn.Linear(dim, 2 * dim, bias=False)
+
+    def interpolate_with_noise(self, z, sampling=False, noise_level=-1):
+        self.diag_gaussian.reset_parameters(self.mean_and_var(z))
+        z_latents = self.diag_gaussian.sample() if sampling else self.diag_gaussian.mean
+
+        if self.training and self.gamma > 0.0:
+            device = z_latents.device
+            s = z_latents.shape
+            batch_size = 1
```
Comment (clessig) [inline src/weathergen/model/parametrised_prob_dist.py]: We can use the code in model.py in l.769 to get the batch size. No one will remember it's there and we should eventually enable batch_size>1 again

File: src/weathergen/train/trainer.py (hunk 16 above)
Comment (clessig) [reply src/weathergen/train/trainer.py:503]: The probabilistic aspect of the autoencoder should learn the data uncertainty (within and across datasets). The forecasting engine should learn the dynamic uncertainty. But let's discuss how to do this best.

File: config/streams/era5_1deg/era5.yml (hunk 20 above)
Comment (sophie-xhonneux) [reply config/streams/era5_1deg/era5.yml]: I got flash attention crashes at some point because it expects at least 8 heads, hence why

Comment (clessig) [reply config/streams/era5_1deg/era5.yml]: I just tested with `num_heads: 4` and it worked. Was the problem on clariden?

Comment (sophie-xhonneux) [reply config/streams/era5_1deg/era5.yml]: it was in some combination of the prediction heads and  flash attn and other things, not sure which anymore, but will revert this

File: src/weathergen/model/parametrised_prob_dist.py (hunk 23 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/parametrised_prob_dist.py]: still given how assimilate_local works at the moment dealing with batch_size > 1 is so outrageously complicated that until we need it I wouldn't bother implementing it

File: src/weathergen/train/trainer.py (hunk 16 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py:503]: discussed synchronously

File: src/weathergen/model/parametrised_prob_dist.py (hunk 22 above)
Comment (sophie-xhonneux) [reply src/weathergen/model/parametrised_prob_dist.py]: Sure
